src/layers/MapboxStyleLayer/MapboxStyleLayer.js

```diff
@@ -204,17 +170,40 @@ class MapboxStyleLayer extends Layer {
       });
   }
 
-  getFeatures2(resolve) {
-    resolve(this.getFeatures());
-  }
+  /**
+   * Apply visibility to style layers that fits the filter function.
+   * @private
+   */
+  // eslint-disable-next-line class-methods-use-this
+  applyLayoutVisibility(mbMap, visible, filterFunc, layerFilter) {
+    const style = mbMap.getStyle();
 
-  setFilter(filter) {
-    const { mbMap } = this.mapboxLayer;
-    this.styleLayers.forEach(({ id }) => {
-      if (mbMap.getLayer(id)) {
-        mbMap.setFilter(id, filter);
+    if (!mbMap || !style) {
+      return;
+    }
+
+    if (filterFunc) {
+      const visibilityValue = visible ? 'visible' : 'none';
+      for (let i = 0; i < style.layers.length; i += 1) {
+        const styleLayer = style.layers[i];
+        if (filterFunc(styleLayer)) {
+          if (mbMap.getLayer(styleLayer.id)) {
+            mbMap.setLayoutProperty(
+              styleLayer.id,
+              'visibility',
+              visibilityValue,
+            );
+            if (layerFilter) {
+              mbMap.setFilter(styleLayer.id, layerFilter);
+            }
```

Comment:
I think it's never used jsut remove this if

Reply:
Removed

Reply:
it's still there :)

Reply:
Done :)